text/0000-intl-v2.md

```diff
@@ -0,0 +1,218 @@
+    @api counter = 0;
+}
+```
+
+We can process both, the template and the fluent files using the same technique, you have the component state (the component instance itself) as the context to render each of these two templates, the HTML file and the fluent file. In this case, both of them can have access to all the data available in the context provided, which in this case is `counter` field, which is a numeric value.
+
+If we render each of these two elements individually, first the fluent, then the template, we can make sure that both have access to the exact same data, in the same tick, and the HTML Template can access to the production of the messages file via the component's `i18n` getter, which can be provided by the framework via the `LightningElement` abstraction.
+
+What do we get out of this solution?
+
+* We pave a way forward into a Labeling solution that can be used both internally and externally.
+* We lift any label logic from the component and into the labels themselves by leveraging the Fluent engine: plurals, variables that can reflect their value into the messages, and other complex use cases that are being handled today by the component authors.
+
+#### LWC Inner-workings
+
+LWC Compiler will have to support fluent, and as customary in LWC, a new convention for filename for those fluent files must be defined. The compilation target of those files must be JS, and there have to be an implicit relationship between files, in this case, any component file can have an associated template file, and an associated fluent file with the messages available to that template.
+
+```
+foo/
+  bar.js
+  bar.html
+  bar.fluent
+```
+
+If a component has more than one template, they all can have access to the same set of messages defined implicitly in the file with the matching name, e.g.:
+
+```
+foo/
+  bar.js
+  bar.fluent
+  x.html
+  y.html
+```
+
+The way LWC can register these association is the same used for the default template, using `registerComponent`. Additionally, we must provide a way for a developer to import fluent dependencies declaratively using import declaration, the same way we allow importing templates. In this case, we must provide a new hook for the developer to return a custom fluent reference, in the same fashion that `render()` method works.
+
+#### Connecting The Dots
+
+When a component is about to be rendered, the engine must use the default associated fluent, if exist. If the hook is defined in user-land, it must be invoked to resolve the custom fluent reference. If the fluent reference is new, an internal fluent instance must be created, binding that instance to the component instance in question. The messages exported by the fluent instance must be made available via `i18n` getter on `LightningElement.prototype`. Each message exported from the fluent instance must be a getter, and must be constructed on demand when the message is accessed. This must be done prior to invocation of the template.
```

Comment:
> If the hook is defined in user-land, it must be invoked to resolve the custom fluent reference.

What does this mean? Are you proposing something like [programmatic style sheets](https://lwc.dev/guide/css#associate-css-stylesheets-to-a-component-programmatically), but for fluent files?

> The messages exported by the fluent instance must be made available via `i18n` getter on `LightningElement.prototype`.

By this, I assume you mean that `this.i18n.foo` works, but `this.template.querySelector('x-component').i18n.foo` does not work, correct? In other words, `i18n` is only exposed to code inside the component, not outside the component?

Reply:
> > If the hook is defined in user-land, it must be invoked to resolve the custom fluent reference.
> 
> What does this mean? Are you proposing something like [programmatic style sheets](https://lwc.dev/guide/css#associate-css-stylesheets-to-a-component-programmatically), but for fluent files?

I'm not sure anymore... if we are saying that this implicit association is sufficiently, then no. Let's think about it.
 
> > The messages exported by the fluent instance must be made available via `i18n` getter on `LightningElement.prototype`.
> 
> By this, I assume you mean that `this.i18n.foo` works, but `this.template.querySelector('x-component').i18n.foo` does not work, correct? In other words, `i18n` is only exposed to code inside the component, not outside the component?

Correct. Just like `template`.